src/operator/semijoin.rs

```diff
@@ -0,0 +1,609 @@
+use crate::{
+    algebra::{HasZero, IndexedZSet, MulByRef, ZSet},
+    circuit::{
+        operator_traits::{BinaryOperator, Data, Operator},
+        Scope,
+    },
+    lattice::Lattice,
+    trace::{ord::OrdValSpine, Batch, BatchReader, Batcher, Builder, Cursor, Trace, TraceReader},
+    Circuit, Stream, Timestamp,
+};
+use deepsize::DeepSizeOf;
+use hashbrown::HashMap;
+use std::{
+    borrow::Cow,
+    cmp::{min, Ordering},
+    fmt::Write,
+    hash::Hash,
+    marker::PhantomData,
+};
+use timely::PartialOrder;
+
```

Comment:
A nice thing about semijoin is that it doesn't take a closure, and hence its results can be cached, so if you semijoin the same streams multiple times in a circuit, no duplicate operators will be created. I can help add caching support to this module.